src/content/app/tools/blast/components/blast-sequence-alignment/blastSequenceAlignmentHelper.ts

```diff
@@ -0,0 +1,165 @@
+  const alignedLines: BlastAlignmentLine[] = [];
+
+  while (cursor < querySequence.length) {
+    const lastAlignedSegment = alignedLines.at(-1);
+    const nextQueryStartPosition = lastAlignedSegment?.queryLineEnd
+      ? lastAlignedSegment?.queryLineEnd + 1 // start next line from the next nucleotide / amino acid
+      : queryStart;
+    const nextHitStartPosition = lastAlignedSegment?.hitLineEnd
+      ? lastAlignedSegment?.hitLineEnd + hitDirection // start nest line from the next nucleotide / amino acid
+      : hitStart;
+
+    const alignedSegment = createAlignedSegment({
+      queryStart: nextQueryStartPosition,
+      hitStart: nextHitStartPosition,
+      hitDirection,
+      querySequence,
+      hitSequence,
+      alignmentLine,
+      cursor
+    });
+
+    alignedLines.push(alignedSegment);
+    cursor += alignmentLineLength;
+  }
+
+  return alignedLines;
+};
+
+const createAlignedSegment = (params: {
+  querySequence: string;
+  hitSequence: string;
+  alignmentLine: string;
+  hitDirection: number; // either 1 or - 1
+  queryStart: number;
+  hitStart: number;
+  alignmentLineLength?: number;
+  cursor: number;
+}): BlastAlignmentLine => {
+  const {
+    querySequence,
+    hitSequence,
+    alignmentLine,
+    hitDirection,
+    queryStart,
+    hitStart,
+    alignmentLineLength = ALIGNMENT_LINE_LENGTH,
+    cursor
+  } = params;
+
+  const startIndex = cursor;
+  const endIndex =
+    querySequence.length - cursor > alignmentLineLength
+      ? cursor + alignmentLineLength - 1
+      : querySequence.length - 1;
+
+  const querySequenceSlice = querySequence.slice(startIndex, endIndex + 1);
+  const alignmentSlice = alignmentLine.slice(startIndex, endIndex + 1);
+  const hitSequenceSlice = hitSequence.slice(startIndex, endIndex + 1);
+
+  const updatedQuerySequenceCounter = calculateEndPosition({
+    sequence: querySequence,
+    startIndex,
+    endIndex,
+    startPosition: queryStart,
+    strandDirectionMultiplier: 1
+  });
+
+  const updatedHitSequenceCounter = calculateEndPosition({
+    sequence: hitSequence,
+    startIndex,
+    endIndex,
+    startPosition: hitStart,
+    strandDirectionMultiplier: hitDirection
+  });
+
+  return {
+    queryLineStart: queryStart,
+    queryLineEnd: updatedQuerySequenceCounter,
+    hitLineStart: hitStart,
+    hitLineEnd: updatedHitSequenceCounter,
+    alignmentLineStart: cursor + 1,
+    alignmentLineEnd: cursor + alignmentLineLength,
+    queryLine: querySequenceSlice,
+    hitLine: hitSequenceSlice,
+    alignmentLine: alignmentSlice
+  };
+};
+
+const calculateEndPosition = ({
+  sequence,
+  startIndex,
+  endIndex,
+  startPosition,
+  strandDirectionMultiplier
+}: {
+  sequence: string;
+  startIndex: number;
+  endIndex: number;
+  startPosition: number;
+  strandDirectionMultiplier: number;
+}) => {
+  let position = startPosition;
+
+  for (let i = startIndex; i < endIndex; i++) {
+    if (sequence[i] !== '-') {
+      position += 1 * strandDirectionMultiplier;
```

Comment:
```suggestion
      position += strandDirectionMultiplier;
```

Reply:
I think I explicitly added the `1` for clarity, i.e. "either add or subtract one based on whether you are moving in the same or in the opposite direction as the query sequence". I realise that `1 * strandDirectionMultiplier` looks silly, because it's the same as just `strandDirectionMultiplier`; but then, will it not be confusing that the code would read as if you are adding a multiplier.

Reply:
Fair enough. Happy to leave it as it is.